rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -256,112 +261,118 @@ private class InMemHotStore[F[_]: Concurrent, C, P, A, K](
-                  (state.copy(joins = state.joins.updated(channel, newVal)), false)
-                }
-              } else (state.copy(joins = state.joins.updated(channel, curJoins)), false)
-            })
+            if (doRemove) {
+              if (outOfBounds)
+                (state.copy(joins = state.joins.updated(channel, curJoins)), true)
+              else {
+                val newVal = removeIndex(curJoins, index)
+                (state.copy(joins = state.joins.updated(channel, newVal)), false)
+              }
+            } else (state.copy(joins = state.joins.updated(channel, curJoins)), false)
+          }
     } yield ()
 
-  def changes(): F[Seq[HotStoreAction]] =
+  def changes: F[Seq[HotStoreAction]] =
     for {
       cache <- hotStoreState.get
-      continuations = (cache.continuations.map {
-        case (k, v) if (v.isEmpty) => DeleteContinuations(k)
-        case (k, v)                => InsertContinuations(k, v)
-      }).toVector
-      data = (cache.data.map {
-        case (k, v) if (v.isEmpty) => DeleteData(k)
-        case (k, v)                => InsertData(k, v)
-      }).toVector
-      joins = (cache.joins.map {
-        case (k, v) if (v.isEmpty) => DeleteJoins(k)
-        case (k, v)                => InsertJoins(k, v)
-      }).toVector
-    } yield (continuations ++ data ++ joins)
+      continuations = cache.continuations.map {
+        case (k, v) if v.isEmpty => DeleteContinuations(k)
+        case (k, v)              => InsertContinuations(k, v)
+      }.toVector
+      data = cache.data.map {
+        case (k, v) if v.isEmpty => DeleteData(k)
+        case (k, v)              => InsertData(k, v)
+      }.toVector
+      joins = cache.joins.map {
+        case (k, v) if v.isEmpty => DeleteJoins(k)
+        case (k, v)              => InsertJoins(k, v)
+      }.toVector
+    } yield continuations ++ data ++ joins
 
   private def removeIndex[E](col: Seq[E], index: Int): Seq[E] = {
     val (l1, l2) = col splitAt index
-    (l1 ++ (l2 tail))
+    l1 ++ (l2 tail)
   }
 
   def toMap: F[Map[Seq[C], Row[P, A, K]]] =
     for {
       cache         <- hotStoreState.get
-      data          = cache.data.map(_.leftMap(Seq(_))).toMap
-      continuations = (cache.continuations ++ cache.installedContinuations.mapValues(Seq(_))).toMap
-      zipped        = zip(data, continuations, Seq.empty[Datum[A]], Seq.empty[WaitingContinuation[P, K]])
-      mapped        = zipped.mapValues { case (d, k) => Row(d, k) }
+      data          = cache.data.map(_.leftMap(Seq(_)))
+      continuations = cache.continuations ++ cache.installedContinuations.mapValues(Seq(_))
+
+      dataAt = (i: Seq[C]) => data.getOrElse(i, Seq.empty[Datum[A]])
+      continuationAt = (i: Seq[C]) =>
+        continuations.getOrElse(i, Seq.empty[WaitingContinuation[P, K]])
+      mapped = (data.keySet ++ continuations.keySet).map { i =>
+        (i, Row(dataAt(i), continuationAt(i)))
+      }.toMap
```

Comment:
Try to rewrite this _zipping_ with the use of `padZip` and `mapValues`, it can little more succinct.

Also remove helper _zip_ function with its tests.

Reply:
@tgrospic, done in [commit](https://github.com/rchain/rchain/pull/3557/commits/d19b2d88dc85e99ceee0bd8ff24c9ef7de0f55f8). Also removed corresponding files.